datastore/filesystem_datastore.nim

```diff
@@ -0,0 +1,159 @@
+import std/os
+
+import pkg/questionable
+import pkg/questionable/results
+import pkg/upraises
+
+import ./datastore
+
+export datastore
+
+push: {.upraises: [].}
+
+type
+  FileSystemDatastore* = ref object of Datastore
+    root: string
+
+const
+  objExt* = ".obj"
+
+proc new*(
+  T: type FileSystemDatastore,
+  root = "data"): ?!T =
+
+  try:
+    let
+      root = if root.isAbsolute: root
+             else: getCurrentDir() / root
+
+    createDir(root)
+    success T(root: root)
+
+  except IOError as e:
+    failure e
+
+  except OSError as e:
+    failure e
+
+proc root*(self: FileSystemDatastore): string =
+  self.root
+
+proc path*(
+  self: FileSystemDatastore,
+  key: Key): string =
+
+  var
+    segments: seq[string]
+
+  for ns in key:
+    without field =? ns.field:
+      segments.add ns.value
+      continue
+
+    segments.add(field / ns.value)
+
+  # is it problematic that per this logic Key(/a:b) evaluates to the same path
+  # as Key(/a/b)? may need to check if/how other Datastore implementations
+  # distinguish them
+
+  self.root / joinPath(segments) & objExt
+
+method contains*(
+  self: FileSystemDatastore,
+  key: Key): ?!bool {.locks: "unknown".} =
+
+  success fileExists(self.path(key))
+
+method delete*(
+  self: FileSystemDatastore,
+  key: Key): ?!void {.locks: "unknown".} =
+
+  let
+    path = self.path(key)
+
+  try:
+    removeFile(path)
+    success()
+
+    # removing an empty directory might lead to surprising behavior depending
```

Comment:
This is extremelly dangerous functionality in a filesystem module in general. We should make sure that this doesn't allow blowing up `/` or some other important directory, perhaps it makes sense to have hardcoded list of common "protected" directories that the user can delete only if setting a flag manually.